draft-irtf-cfrg-opaque.md

```diff
@@ -1145,33 +1149,75 @@ gets the most significant bytes of the HKDF output, Km2 the next bytes, etc.
 
 Values IKM and info are defined for each protocol:
 
-#### HMQV key derivation
+#### HMQV key derivation {#derive-3dh}
+
+The HKDF input parameter `info` is computed as follows:
+
+~~~
+info = "HMQV keys" || len(nonceU) || nonceU
+                   || len(nonceS) || nonceS
+                   || len (IdU) || IdU
+                   || len(IdS) || IdS
+~~~
+
+The input parameter `IKM` is `Khmqv`, where `Khmqv` is computed by the client as follows:
+
+~~~
+1. u' = (eskU + u\*skU) mod p
+2. Khmqv = (epkS \* pkS^s)^u'
+~~~
 
- - info = "HMQV keys" \| nonceU \| nonceS \| IdU \| IdS
+where `u` is computed as follows:
 
- - IKM = Khmqv
+~~~
+hashInput = len(epkU) || epkU ||
+            len(info) || info ||
+            len("client") || "client"
+u = Hash(hashInput) mod p
+~~~
 
-   where Khmqv is computed:
+Hash is the same hash function used in the main OPAQUE protocol for key derivation.
+Its output length must be at least the length of the group order p.
 
-   - by the client:  Khmqv = (epkS \* pkS^s)^{eskU + u\*skU}
+Likewise, servers compute `Khmqv` as follows:
 
-   - by the server:  Khmqv = (epkU \* pkU^u)^{eskS + u\*skS}
+~~~
+1. s' = (eskS + s\*skS) mod p
+2. Khmqv = (epkU \* pkU^u)^s'
+~~~
 
-   and u = H(epkU \| "client" \| info) and s = H(epkS \| "server" \| info).
+where `s` is computed as follows:
 
-<!--     (u and s are often denoted d and e)   -->
+~~~
+hashInput = len(epkS) || epkS ||
+            len(info) || info ||
+            len("server") || "client"
+s = Hash(hashInput) mod p
+~~~
 
 #### 3DH key derivation
 
- - info = "3DH keys" \| nonceU \| nonceS \| IdU \| IdS
+The HKDF input parameter `info` is computed as follows:
```

Comment:
remove last line

Reply:
Which line?

Reply:
(I wasn't sure what line you were referring to, so I'll merge now and can clean this up later!)

Reply:
My bad. I read HMQV instead of HKDF (too many H's :-)